### ISSUE TITLE
Include laws for Functor, Applicative, Monad

### ISSUE THREAD

**Issue by alex-mckenna** [agda/agda-stdlib#496]
Currently, the modules for Functor, Applicative and Monad include records for the Raw types similar to the algebraic structures, but do not have corresponding types including the laws. I think this could be a nice issue for someone like myself who knows the laws, and wants to take a gentle first step into contributing to the standard library.

From what I can tell this would involve:

  - [ ] Creating a `Category.Structures` module for definitions of `IsFunctor`, `IsApplicative` and `IsMonad`
  - [ ] Creating the extended records which include proof the laws hold

**Comment by alex-mckenna** [comment]
@gallais Any comments / gotchas I should know about with this?

**Comment by gallais** [comment]
> Any comments / gotchas I should know about with this?

The right notion of equality to use in these is a hard question. E.g. you'll probably need
functional extensionality to prove the laws for `State` if you pick `PropositionalEquality`
as your notion of equality.

**Comment by HuStmpHrrr** [comment]
i think this is a good issue. I was wondering the same thing too.

**Comment by jespercockx** [comment]
It would be great if someone could take a look at this! It is kind of embarrassing to claim one of the big advantages of Agda over Haskell is the ability to specify lawful typeclasses, but then the standard library does not actually do this.

**Comment by JacquesCarette** [comment]
I was hoping to start migrating the basic parts of `agda-categories` to `agda-stdlib`, so that `Category`, `Functor` and friends could be 'right'.  I think that things in `agda-categories` have settled, so that once stdlib-1.3 ships, that could start.

**Comment by jamesmckinna** [comment]
... fast-forward to v2.x and year 2024...

**Comment by JacquesCarette** [comment]
Right, this serves as a nice reminder for me to start a design discussion on just that.
